feat: add hidden skip-to-content button

diivi opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #47523 

Working:
![skip to content button in action](https://user-images.githubusercontent.com/41837037/191487868-648269ff-36ba-4ba9-b39a-ba06f1fe2691.gif)

diivi committed:
feat: add skip-to-content button

diivi commented:
This will require an id of #content-start on the first heading/ result of the news page too (search results). I was asked to mention this here, for more info - https://discord.com/channels/692816967895220344/715074489422970962/1022137562271330314

ojeytonwilliams reviewed (changes requested):
HI @diivi thanks for putting this together!

As I mentioned in discord, there's no need to create multiple ids to jump to.  The default layout wraps every page, so you can jump to an id inside that.

If we go this route we basically guarantee an inconsistent experience. Future maintainers are likely to forget to add the `id='content-start'` to a new type of page.

diivi committed:
fix: remove content start ids from everywhere and add to default layout

diivi committed:
chore: format landing-top.tsx

diivi committed:
use single quotes in skip-to-content component

ojeytonwilliams reviewed (dismissed):
Thanks, @diivi, this works really well. LGTM :+1:

bbsmooth commented:
> there's no need to create multiple ids to jump to

@ojeytonwilliams I'm going to disagree a little here. For the Steps, the skip link should focus on the `h1` in the instructions. If it focuses on the `#content-start` div that wraps `main` instead, then the user will still have to tab through the bread crumb links and editor/console/preview buttons. 

My suggestion is to add an `id` to the `h1` and have the skip link jump to that.

ojeytonwilliams commented:
Thanks @bbsmooth, good catch! I agree that it should skip all the navigation, not some of it.  However, the root of the problem is that the navigation components are scattered across the site.

To keep things DRY and reasonably easy to understand, I think it would be better to pull the BreadCrumbs up to the `default-layout` rather than pushing the target ids down into 8 components.  This may require a bit of Gatsby magic, but it should be doable.

diivi commented:
I don't think moving `ActionRow` from `DesktopLayout` to `DefaultLayout` will be that straightforward because of all the props that it requires, I even tried doing so. 
Could you point me in the right direction explaining how I can pull the `ActionRow` component out of `DesktopLayout` component?
![desktoplayout.tsx](https://user-images.githubusercontent.com/41837037/191582123-fb4ac2a3-fbb3-4866-82b0-d3143ca5f9ac.png)

ojeytonwilliams commented:
Sure. The Gatsby [docs](https://www.gatsbyjs.com/docs/programmatically-create-pages-from-data/#specifying-a-template) are pretty good, but the main thing you need to do is make the data available inside `layout-selector.tsx`.  That can be achieved via a GraphQL query which Gatsby will run and then inject the result as a `data` prop.

So, I'd use the approach in client/src/templates/Challenges/classic/show.tsx as a guide.

Finally, there's no need to move the entire ActionRow, just the BreadCrumb part.  Reason being, the goal is to put all the navigation elements in the DefaultLayout, but the BreadCrumb is the only part of ActionRow that deals with navigation.

I hope that helps!

bbsmooth commented:
> there's no need to move the entire ActionRow

I personally think that the skip to content link should skip both the bread crumbs and the buttons in the ActionRow, but there is no hard rule on where it should skip to. If we do feel the buttons are part of the main content and shouldn't be skipped then I suggest we might change the text on the skip link to "Skip navigation" which I think would more accurately reflect what the link is doing here.

diivi committed:
include breadcrumbs in navigation

diivi commented:
I elevated the main breadcrumb navigation, it's now a part of the default layout component.
![fcc](https://user-images.githubusercontent.com/41837037/191934844-3af72337-5fcb-458b-81e2-79b815db856b.gif)

diivi committed:
linting fail fix

ojeytonwilliams commented:
@diivi this is looking great, really nice work.  The only issue I'm running into is that if I tab backwards after skipping past the breadcrumbs, the breadcrumbs receive focus as expected, but they don't scroll into view.  Other pages like /learn, don't have this problem - the page scrolls as I tab around.

Having said all that, this is already present on main, so we've not made things worse.  As such, I'm happy to merge this is if @bbsmooth is and we can deal with the lack of scrolling another time.

Sboonny commented on client/src/components/Header/index.tsx:
```diff
@@ -87,6 +88,7 @@ export class Header extends React.Component<
           <style>{':root{--header-height: 38px}'}</style>
         </Helmet>
         <header>
+          <SkipToContentButton />
```
there is no need to DRY the button, it should be used only once and if it's used multiple times in the future, then we need to change the structure of the page instead of DRY the button.

```suggestion
<a href='#content-start' className='skip-to-content-button'>Skip To Content</a>
```

Sboonny commented on client/src/components/Header/index.tsx:
```diff
@@ -6,6 +6,7 @@ import React from 'react';
 import Helmet from 'react-helmet';
 
 import UniversalNav from './components/universal-nav';
+import SkipToContentButton from './components/skip-to-content-button';
```
```suggestion
```

Sboonny commented on client/src/components/Header/components/skip-to-content-button.tsx:
```diff
@@ -0,0 +1,24 @@
+import React, { useState } from 'react';
+import './skip-to-content-button.css';
+
+const SkipToContentButton = (): JSX.Element => {
+  const [showSkipButton, setShowSkipButton] = useState(false);
+
+  return (
+    <div
+      className={`skip-to-content-button ${
+        showSkipButton ? 'show-skip-button' : 'hide-skip-button'
+      }`}
+    >
+      <a
+        href={'#content-start'}
+        onFocus={() => setShowSkipButton(true)}
+        onBlur={() => setShowSkipButton(false)}
+      >
+        Skip Navigation
+      </a>
+    </div>
+  );
+};
+
+export default SkipToContentButton;
```
You can delete this component and move the anchor to `index.tsx`, reason:

- using JavaScript for styling is dangerous because if JavaScript break the whole page break, while the site can function with broken CSS fine

- it's better for the performance to use CSS

```suggestion
```

Sboonny commented on client/src/components/Header/components/skip-to-content-button.css:
```diff
@@ -0,0 +1,10 @@
+.skip-to-content-button {
+  position: absolute;
+  top: var(--header-height);
+}
+.skip-to-content-button.hide-skip-button {
+  left: -1000px;
+}
+.skip-to-content-button.show-skip-button {
+  left: 0px;
+}
```
the `index.tsx` page is using `header.css`, so there is no need to create new file

```suggestion
```

if you decide to move it, could you change the styles too?

You can add some basic styles for now, maybe I will discuss with ahmad how to style it, but that is for different PR.

```css
.skip-to-content-button {
  position: absolute;
  top: var(--header-height);
  border-radius: 15px;
  font-weight: 600;
  padding-block: 1em;
  padding-inline: 1.5em;
  left: -1000px;
}

.skip-to-content-button:focus {
  left: 0px;
}
```

Sboonny reviewed (changes requested):
Thank you for the awesome work, here are some nitpicks and some refactoring

ojeytonwilliams commented:
@ahmadabdolsaheb as discussed in Discord, these code changes result in duplicate breadcrumbs on all but the new project-based curriculum.  Like so:

![image](https://user-images.githubusercontent.com/15801806/192008119-649ca5b3-2354-42a8-826f-995aada5af46.png)

and

![image](https://user-images.githubusercontent.com/15801806/192008450-5da5ac69-082b-4d26-bc92-fcbd73a55fb4.png)


Would it be okay if simply deleted the smaller ones? That way it's a consistent experience across all our challenge pages.

diivi commented:
Waiting for @ahmadabdolsaheb to comment on this to continue.

ahmadabdolsaheb commented:
Yes, the smaller breadcrumb can go.

diivi committed:
use anchor tag instead of new component

Sboonny commented:
@diivi I am wondering, is the skip navigation changes related to breadcrumb changes? 

Because it's hard to distinguished "why the test are failing", if it's possible to create two PRs, so reviewing and fixing can be easier that would be awesome 👍 . Saying that, maybe I am wrong and breadcrumb is necessary, after all I only skimmed through breadcrumbs discussion

here is what I found reading cypress errors

- the new `breadcrumbs` are stacking on top of `textarea`
- the new elements have changed how the old elements structured (html structure), if people are okay with the new hierarchy, running `npm run test -- -u` or `npx jest -u` should create a new snapshot  

I have no idea why localization test fails 🤷, I just hope fixing the other two will clear it as well.

diivi commented:
I was asked to make the breadcrumbs part of the navigation and not the main content. So I moved them. But then I noticed, on some pages, there were two breadcrumbs, so I removed the second ones with this change. This makes skip to content skip the breadcrumbs too, which shouldn't be a part of the main content but only the navigation bar/header. This is why the two changes are part of the same PR, because if I just add skip to content, it will skip to the breadcrumbs. Should I revert to the point where I just added the skip to content button and skip the breadcrumb navigation in another PR?

naomi-lgbt commented:
No, IMHO the removal of the old breadcrumbs is within the scope of this PR and should stay.

Sboonny commented:
@diivi thank you for more context, it looks like related to skip to the button, so please keep it one PR. sorry for the confusion 👍

Please run one of the snapshot generating command, and fix the first issue however you see fit.

Although I have already seen you in discord and there is no need to link this again, but here it's anyway

If you need help, please join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3).

diivi committed:
update challenge title snap

ojeytonwilliams committed:
fix(test): reliably move focus onto the editor

ojeytonwilliams reviewed (approved):
LGTM :+1: 

Nice work, @diivi!

bbsmooth commented:
This works really well. LGTM :+1:

Sboonny commented:
Thank you for the awesome work
